pandas/core/reshape/merge.py

```diff
@@ -1089,13 +1089,15 @@ def _validate_specification(self):
                     raise ValueError('len(left_on) must equal the number '
                                      'of levels in the index of "right"')
                 self.right_on = [None] * n
+            self.right_on = self.right_on or [None] * (n + 1)
```

Comment:
why do you need to add 1 this seems odd

Reply:
so, the thinking here was to not add additional check and force the cases where either of left_on or right_on is absent to the if clause at 1099.

Reply:
right

but i think this is actually changing things slightly

maybe let’s go back to your original patch

Reply:
> but i think this is actually changing things slightly

additional message checks maybe should be added to the following tests as a precursor. xref #23922

```
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(MergeError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(ValueError):
pandas/tests/reshape/merge/test_merge_asof.py:        with pytest.raises(ValueError):
pandas/tests/reshape/merge/test_multi.py:        with pytest.raises(ValueError):
pandas/tests/reshape/merge/test_multi.py:        with pytest.raises(ValueError):
```

also the following tests perhaps should be moved to pandas/tests/reshape/merge/test_merge.py ( could potentially add additional parametrisation to these tests rather than creating new)

https://github.com/pandas-dev/pandas/blob/a7382231fe2512c43dce7ab56975b2cf4f239e9d/pandas/tests/reshape/merge/test_join.py#L202-L247